internal/pkg/storageos/client_test.go

```diff
@@ -0,0 +1,199 @@
+package storageos
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/golang/mock/gomock"
+	api "github.com/storageos/go-api/v2"
+
+	"github.com/storageos/api-manager/internal/pkg/storageos/mocks"
+)
+
+func TestClient_Refresh(t *testing.T) {
+	secretPath, err := ioutil.TempDir("", "secrets")
+	if err != nil {
+		t.Fatalf("failed to create temp dir: %v", err)
+	}
+	defer os.RemoveAll(secretPath)
+	if err := ioutil.WriteFile(filepath.Join(secretPath, "username"), []byte("username"), 0644); err != nil {
+		t.Fatalf("failed to write username file: %v", err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(secretPath, "password"), []byte("password"), 0644); err != nil {
+		t.Fatalf("failed to write password file: %v", err)
+	}
+
+	token := "faketoken"
+
+	errRefresh := errors.New("refresh error")
+	errAuth := errors.New("auth error")
+
+	tests := []struct {
+		name       string
+		ctx        context.Context
+		interval   time.Duration
+		timeout    time.Duration
+		resets     int
+		expect     func(ctx context.Context, m *mocks.MockControlPlane)
+		wantErrors map[error]int
+	}{
+		{
+			name:     "standard refresh",
+			ctx:      context.WithValue(context.Background(), api.ContextAccessToken, token),
+			interval: 100 * time.Millisecond,
+			timeout:  1050 * time.Millisecond,
+			expect: func(ctx context.Context, m *mocks.MockControlPlane) {
+				httpResp := &http.Response{
+					Header: http.Header{
+						"Authorization": []string{"Bearer " + token},
+					},
+					Body: ioutil.NopCloser(bytes.NewReader([]byte(""))),
+				}
+				m.EXPECT().RefreshJwt(gomock.Any()).Return(api.UserSession{}, httpResp, nil).Times(10)
+			},
+		},
+		{
+			name:     "recovery from refresh error",
+			ctx:      context.WithValue(context.Background(), api.ContextAccessToken, token),
+			interval: 100 * time.Millisecond,
+			timeout:  1050 * time.Millisecond,
+			expect: func(ctx context.Context, m *mocks.MockControlPlane) {
+				httpResp := &http.Response{
+					Header: http.Header{
+						"Authorization": []string{"Bearer newtoken"},
+					},
+					Body: ioutil.NopCloser(bytes.NewReader([]byte(""))),
+				}
+				m.EXPECT().RefreshJwt(gomock.Any()).Return(api.UserSession{}, nil, errRefresh).Times(1)
+				m.EXPECT().AuthenticateUser(gomock.Any(), api.AuthUserData{Username: "username", Password: "password"}).Return(api.UserSession{}, httpResp, nil).Times(1)
```

Comment:
Let's use constants for all the "username" and "password" literals in all the test cases, reducing the chances of typo based failures.